go-controller/pkg/clustermanager/endpointslicemirror/endpointslice_mirror_controller.go

```diff
@@ -0,0 +1,428 @@
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	var mirroredEndpointSlice *v1.EndpointSlice
+
+	slices, err := c.endpointSliceLister.EndpointSlices(namespace).List(mirrorEndpointSliceSelector)
+	if err != nil {
+		return err
+	}
+
+	if len(slices) == 1 {
+		mirroredEndpointSlice = slices[0]
+	}
+	if len(slices) > 1 {
+		klog.Errorf("Found %d mirrored EndpointSlices for %s/%s, removing all of them", len(slices), namespace, name)
+		if err := c.kubeClient.DiscoveryV1().EndpointSlices(namespace).DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: mirrorEndpointSliceSelector.String()}); err != nil {
+			return err
+		}
+		// return an error so there is a retry that will recreate the correct mirrored EndpointSlice
+		return fmt.Errorf("found and removed %d mirrored EndpointSlices for %s/%s", len(slices), namespace, name)
+	}
+
+	if defaultEndpointSlice == nil {
+		if mirroredEndpointSlice != nil {
+			klog.Infof("The default EndpointSlice %s/%s no longer exists, removing the mirrored one: %s", namespace, mirroredEndpointSlice.Labels[LabelSourceEndpointSlice], cache.MetaObjectToName(mirroredEndpointSlice))
+			return c.kubeClient.DiscoveryV1().EndpointSlices(namespace).Delete(ctx, mirroredEndpointSlice.Name, metav1.DeleteOptions{})
+		}
+		klog.Infof("The default EndpointSlice %s/%s no longer exists", namespace, name)
+		return nil
+	}
+
+	// defaultEndpointSlice cannot be nil beyond this point
+	if defaultEndpointSlice.AddressType == v1.AddressTypeFQDN {
+		if mirroredEndpointSlice != nil {
+			klog.Infof("The default EndpointSlice %s is of type %q, removing the mirrored one: %s", cache.MetaObjectToName(defaultEndpointSlice), v1.AddressTypeFQDN, cache.MetaObjectToName(mirroredEndpointSlice))
+			return c.kubeClient.DiscoveryV1().EndpointSlices(namespace).Delete(ctx, mirroredEndpointSlice.Name, metav1.DeleteOptions{})
+		}
+		return nil
+	}
+
+	if mirroredEndpointSlice != nil {
+		// nothing to do if we already reconciled this exact EndpointSlice
+		if mirroredResourceVersion, ok := mirroredEndpointSlice.Labels[LabelSourceEndpointSliceVersion]; ok {
+			if mirroredResourceVersion == defaultEndpointSlice.ResourceVersion {
+				return nil
+			}
+		}
+	}
+
+	currentMirror := c.mirroredEndpointSlice(mirroredEndpointSlice, defaultEndpointSlice, namespacePrimaryNetwork.Name)
+	isIPv6 := defaultEndpointSlice.AddressType == v1.AddressTypeIPv6
+	for i, endpoint := range defaultEndpointSlice.Endpoints {
+		if endpoint.TargetRef != nil && endpoint.TargetRef.Kind == "Pod" {
+			podIP, err := c.getPodIP(endpoint.TargetRef.Name, endpoint.TargetRef.Namespace, namespacePrimaryNetwork.NADName, isIPv6)
+			if err != nil {
+				return fmt.Errorf("failed to determine the Pod IP of: %s/%s: %v", endpoint.TargetRef.Namespace, endpoint.TargetRef.Name, err)
+			}
+			newEp := endpoint.DeepCopy()
+			newEp.Addresses = []string{podIP}
+			currentMirror.Endpoints[i] = *newEp
+		}
+	}
```

Comment:
why not do this as well in the mirrorEndpointSlice function?

Reply:
Good point, done.